include/beast/websocket/impl/write.ipp

```diff
@@ -244,101 +168,335 @@ operator()(error_code ec, bool again)
+                }
+                else
+                {
+                    BOOST_ASSERT(d.ws.wr_.buf_size != 0);
+                    d.remain = buffer_size(d.cb);
+                    if(d.remain > d.ws.wr_.buf_size)
+                        d.entry = do_nomask_frag;
+                    else
+                        d.entry = do_nomask_nofrag;
+                }
+            }
+            else
+            {
+                if(! d.ws.wr_.autofrag)
+                {
+                    d.entry = do_mask_nofrag;
+                }
+                else
+                {
+                    BOOST_ASSERT(d.ws.wr_.buf_size != 0);
+                    d.remain = buffer_size(d.cb);
+                    if(d.remain > d.ws.wr_.buf_size)
+                        d.entry = do_mask_frag;
+                    else
+                        d.entry = do_mask_nofrag;
+                }
+            }
+            d.state = do_maybe_suspend;
+            break;
 
-        case 1:
+        //----------------------------------------------------------------------
+
+        case do_nomask_nofrag:
         {
-            if(! d.fh.mask)
-            {
-                // send header and entire payload
-                d.state = 99;
-                BOOST_ASSERT(! d.ws.wr_block_);
-                d.ws.wr_block_ = &d;
-                boost::asio::async_write(d.ws.stream_,
-                    buffer_cat(d.fh_buf.data(), d.cb),
+            d.fh.fin = d.fin;
+            d.fh.len = buffer_size(d.cb);
+            detail::write<static_streambuf>(
+                d.fh_buf, d.fh);
+            // Send frame
+            d.state = do_upcall;
+            BOOST_ASSERT(! d.ws.wr_block_);
+            d.ws.wr_block_ = &d;
+            boost::asio::async_write(d.ws.stream_,
+                buffer_cat(d.fh_buf.data(), d.cb),
+                    std::move(*this));
+            return;
+        }
+
+        //----------------------------------------------------------------------
+
+        case do_nomask_frag:
+        {
+            auto const n = clamp(
+                d.remain, d.ws.wr_.buf_size);
+            d.remain -= n;
+            d.fh.len = n;
+            d.fh.fin = d.fin ? d.remain == 0 : false;
+            detail::write<static_streambuf>(
+                d.fh_buf, d.fh);
+            // Send frame
+            d.state = d.remain == 0 ?
+                do_upcall : do_nomask_frag + 1;
+            BOOST_ASSERT(! d.ws.wr_block_);
+            d.ws.wr_block_ = &d;
+            boost::asio::async_write(d.ws.stream_,
+                buffer_cat(d.fh_buf.data(),
+                    prepare_buffers(n, d.cb)),
                         std::move(*this));
+            return;
+        }
+
+        case do_nomask_frag + 1:
+            d.cb.consume(
+                bytes_transferred - d.fh_buf.size());
+            d.fh_buf.reset();
+            d.fh.op = opcode::cont;
+            if(d.ws.wr_block_ == &d)
+                d.ws.wr_block_ = nullptr;
```

Comment:
More a matter of style than anything, in this code I am very careful to always copy and paste idioms so there is no variation.